metaconfig-core/shared/src/main/scala/metaconfig/generic/package.scala

```diff
@@ -7,8 +7,12 @@ package object generic {
     macro metaconfig.internal.Macros.deriveSurfaceImpl[T]
   def deriveDecoder[T](default: T): ConfDecoder[T] =
     macro metaconfig.internal.Macros.deriveConfDecoderImpl[T]
+  def deriveDecoderFactory[T]: ConfDecoderFactory[T] =
+    macro metaconfig.internal.Macros.deriveConfDecoderFactoryImpl[T]
   def deriveEncoder[T]: ConfEncoder[T] =
     macro metaconfig.internal.Macros.deriveConfEncoderImpl[T]
   def deriveCodec[T](default: T): ConfCodec[T] =
     macro metaconfig.internal.Macros.deriveConfCodecImpl[T]
+  def deriveCodecFactory[T]: ConfCodecFactory[T] =
```

Comment:
Commenting the review suggestion inline to elaborate on what I mean. Would it be possible to do something like this instead? The other decoder/encoder macros would then be implemented on top of this macro expansion. It would fix https://github.com/scalameta/metaconfig/issues/58 if we make the `default` argument by-name
```suggestion
  def deriveCodecFactory[T]: T => ConfCodec[T] =
```

Reply:
this, or something very similar to this, is what didn't work in my first patch, leading to a circular dependency in one of the files which i worked around (without really understanding why it was failing) by moving some code into the companion object (and you requested to move back).

your suggestion will make one such usage simpler (but perhaps it could be emulated with an implicit conversion), but it would make chaining harder. what i am trying to allow, for instance, is the `scalafmt` case of presets (and also, if possible, avoiding creating an instance of a decoder for each instance of the class).

this is what `scalafmt` does today (all demonstrative, not exact):
```
class ScalafmtConfig(...) {
  implicit val decoderField1 = decoderForField1(field1)
  ...
  implicit val decoderFieldN = decoderForFieldN(fieldN)
  implicit val decoderThis = generic.deriveDecoder(this)
}
```
and then for presets I needed to modify it like this:
```
class ScalafmtConfig(...) {
  implicit val decoderField1 = decoderForField1(field1) // hopefully we don't forget presets here
  ...
  implicit val decoderFieldN = decoderForFieldN(fieldN)
  private val baseDecoderThis = generic.deriveDecoder(this)
  implicit val decoderThis = decoderForPresets(baseDecoderThis)
}
```
and this is what I would like to accomplish:
```
class Align(...)
object Align {
  implicit val decoder = ... // includes presets and possibly tokens.add handling
}
...
class ScalafmtConfig(...) {
 ...
}
object ScalafmtConfig {
  private val baseDecoder = generic.deriveDecoder(ScalafmtConfig.default)
  implicit val decoder = decoderForPresets(baseDecoderThis)
}
```
meaning, the decoder for a type will be defined with the type and every class using it will get it automatically. (and one instance, too.)

for that, special-case handling will need to be implemented as chaining of methods (or factories). presets handling looks like decoding `read(conf: Conf): Configured[(A, Conf)]` and then using that to apply regular generic decoder `read(conf: Conf, A: preset)`. and that needs to happen for each field, too, so must be inside `Conf.getSettingOrElse`.

hopefully, this makes some sense for you to comment further.